chainimport/headers_import.go

```diff
@@ -0,0 +1,1919 @@
+	default:
+		return UnknownHeaderSize
+	}
+}
+
+// HeaderMetadata contains the metadata about the header source.
+type HeaderMetadata struct {
+	BitcoinChainType BitcoinChainType
+	HeaderType       HeaderType
+	StartHeight      uint32
+	EndHeight        uint32
+}
+
+// HeaderImportSource is a generic interface for loading headers of type T.
+type HeaderImportSource[T any] interface {
+	Open() error
+	Close() error
+	GetHeaderMetadata() (*HeaderMetadata, error)
+	Iterator(start, end int) HeaderIterator[T]
+	GetHeader(index int) (T, error)
+	HeadersCount() int
+	SourceName() string
+}
+
+// FileHeaderSource implements HeaderSource for files.
+type FileHeaderImportSource[T any] struct {
+	Path     string
+	reader   *mmap.ReaderAt
+	fileSize int
+	metadata *HeaderMetadata
+}
+
+// Open opens the file and initializes the mmap reader.
+func (f *FileHeaderImportSource[T]) Open() error {
+	// Only open if not already open
+	if f.reader != nil {
+		return nil
+	}
+
+	reader, err := mmap.Open(f.Path)
+	if err != nil {
+		return fmt.Errorf("failed to mmap file: %w", err)
+	}
+	f.reader = reader
+	f.fileSize = reader.Len()
+
+	// Read and initialize metadata from the beginning of the file.
+	metadata, err := f.GetHeaderMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to get header metadata: %w", err)
+	}
+
+	// Validate that the generic type T matches the header type in the file.
+	if err := f.validateTypeMatch(metadata); err != nil {
+		return fmt.Errorf("type validation failed: %w", err)
+	}
+	f.metadata = metadata
+	f.metadata.EndHeight = metadata.StartHeight + uint32(f.HeadersCount()) - 1
+
+	return nil
+}
+
+func (f *FileHeaderImportSource[T]) Close() error {
+	return f.reader.Close()
+}
+
+// GetHeaderMetadata reads the metadata from the file.
+func (f *FileHeaderImportSource[T]) GetHeaderMetadata() (*HeaderMetadata, error) {
+	if f.metadata != nil {
+		return f.metadata, nil
+	}
+
+	if f.reader == nil {
+		return nil, errors.New("file reader not initialized")
+	}
+
+	// Read the first 6 bytes containing metadata.
+	buf := make([]byte, HeaderMetadataSize)
+	_, err := f.reader.ReadAt(buf, 0)
+	if err != nil {
+		return &HeaderMetadata{}, fmt.Errorf("failed to read metadata: %w", err)
+	}
+
+	// Parse metadata.
+	metadata := HeaderMetadata{
+		BitcoinChainType: BitcoinChainType(buf[0]),
+		HeaderType:       HeaderType(buf[1]),
+		StartHeight:      binary.LittleEndian.Uint32(buf[2:6]),
+	}
+
+	// Validate header type.
+	switch metadata.HeaderType {
+	case BlockHeader, FilterHeader:
+	default:
+		metadata = HeaderMetadata{}
+		err = fmt.Errorf("invalid header type: %s", metadata.HeaderType)
+	}
+
+	return &metadata, err
+}
+
+// validateTypeMatch ensures that the generic type T matches the header type in
+// the file.
+func (f *FileHeaderImportSource[T]) validateTypeMatch(
+	metadata *HeaderMetadata) error {
+
+	var empty T
+	switch any(empty).(type) {
+	case headerfs.BlockHeader:
+		if metadata.HeaderType != BlockHeader {
```

Comment:
This could be done with a `ValidateMetadata(m *HeaderMetadata)` method on both header types that makes sure the correct metadata is provided.